File: typescript/src/kahnQueue/nodeMachine.ts
import { StateMachine } from "../utils/stateMachine.js";

export type NodeState = "QUEUED" | "READY" | "ACTIVE" | "COMPLETE" | "PRUNED";

const NODE_TRANSITIONS: Partial<Record<NodeState, NodeState[]>> = {
  QUEUED: ["READY", "PRUNED"],
  READY: ["ACTIVE", "PRUNED"],
  ACTIVE: ["COMPLETE", "PRUNED"],
  COMPLETE: [],
  PRUNED: [],
};

export class NodeMachine extends StateMachine<NodeState> {
  numSources: number;
  readonly id: number;

  private constructor(initialState: NodeState, id: number, numSources: number) {
    super(initialState, NODE_TRANSITIONS);
    this.id = id;
    this.numSources = numSources;
  }

  static create(id: number, numSources: number): NodeMachine {
    const m = new NodeMachine("QUEUED", id, numSources);
    m.tryReady();
    return m;
  }

  override canTransition(to: NodeState): boolean {
    if (this.is("QUEUED") && this.numSources > 0 && to === "READY") {
      return false;
    }
    return super.canTransition(to);
  }

  decrement(): void {
    if (this.numSources === 0) {
      throw new Error("Attempting to decrement below zero");
    }
    this.numSources--;
    this.tryReady();
  }

  tryReady(): void {
    if (this.canTransition("READY")) {
      this.transition("READY");
    }
  }
}

File: python/src/kahnQueue/node_machine.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Final, Mapping, Set

from kahnQueue.node_state import NodeState
from utils.state_machine import StateMachine

NODE_TRANSITIONS: Final[Mapping[NodeState, Set[NodeState]]] = {
    NodeState.QUEUED: {NodeState.READY, NodeState.PRUNED},
    NodeState.READY: {NodeState.ACTIVE, NodeState.PRUNED},
    NodeState.ACTIVE: {NodeState.COMPLETE, NodeState.PRUNED},
    NodeState.COMPLETE: set(),
    NodeState.PRUNED: set(),
}


@dataclass
class NodeMachine(StateMachine[NodeState]):
    num_sources: int
    id: int

    @classmethod
    def create(cls, id: int, num_sources: int) -> NodeMachine:
        """Create a new NodeMachine and attempt initial transition to READY."""
        m = cls(
            _state=NodeState.QUEUED,
            _transitions=NODE_TRANSITIONS,
            num_sources=num_sources,
            id=id,
        )
        m._try_ready()
        return m

    def can_transition(self, to: NodeState) -> bool:
        if self.is_(NodeState.QUEUED) and self.num_sources > 0 and to == NodeState.READY:
            return False
        return super().can_transition(to)

    def decrement(self) -> None:
        """Decrement num_sources and transition to READY if possible."""
        if self.num_sources <= 0:
            raise RuntimeError("Attempting to decrement below zero")
        self.num_sources -= 1
        self._try_ready()

    def _try_ready(self) -> None:
        if self.can_transition(NodeState.READY):
            self.transition(NodeState.READY)
            